sql/core/src/main/scala/org/apache/spark/sql/execution/exchange/ExchangeCoordinator.scala

```diff
@@ -83,16 +83,17 @@ import org.apache.spark.sql.execution.{ShuffledRowRDD, SparkPlan}
  *  - post-shuffle partition 3: pre-shuffle partition 3 and 4 (size 50 MB)
  */
 class ExchangeCoordinator(
-    numExchanges: Int,
     advisoryTargetPostShuffleInputSize: Long,
     minNumPostShufflePartitions: Option[Int] = None)
   extends Logging {
 
   // The registered Exchange operators.
   private[this] val exchanges = ArrayBuffer[ShuffleExchangeExec]()
 
+  private[this] lazy val numExchanges = exchanges.size
```

Comment:
can we make it a `def`?

Reply:
ya, we can do so, But, I used `lazy val` there because IMO that made us easily notice some bugs about an illegal method call order. For example, all the exchange should be registered before `ExchangeCoordinator.postShuffleRDD` called first time. If a new exchange is wrongly registered after  `ExchangeCoordinator.postShuffleRDD` called, the assertion fails. WDYT?

Reply:
makes sense, let's document it.

Reply:
ok, I'll update